src/hotspot/share/opto/macro.cpp

```diff
@@ -853,6 +853,12 @@ SafePointScalarObjectNode* PhaseMacroExpand::create_scalarized_object_descriptio
   sobj->init_req(0, C->root());
   transform_later(sobj);
 
+  if (iklass && iklass->is_inlinetype()) {
+    // Value object has two additional inputs before the non-static fields
+    sfpt->add_req(_igvn.intcon(1));
+    sfpt->add_req(_igvn.intcon(alloc->_larval ? 1 : 0));
+  }
```

Comment:
This issue happens when scalarizing an merging allocation which has inlinetype, instead of scalarizing a single `InlineTypeNode`. Skiping over doing this optimization for InlineTypeNode may lose some optimization oppportunity? @TobiHartmann , could you please help check this change or give any suggestion on that? Thanks so much!

Reply:
> This issue happens when scalarizing an merging allocation which has inlinetype, instead of scalarizing a single `InlineTypeNode`. Skiping over doing this optimization for InlineTypeNode may lose some optimization oppportunity? 

Thanks @XiaohongGong , IIRC we do allocate a buffer for an InlineTypeNode **if and only iff** it has a non-flattened user, this mean allocated buffers are non-eliminatable, and regular escape analysis will not able to remove allocations and perform scalar replacements. In principle we do need to pass larval state for value object reconstruction during de-optimization, so additional booking for larval state is correct.